controller/volume_controller.go

```diff
@@ -1315,29 +1322,61 @@ func (vc *VolumeController) replenishReplicas(v *longhorn.Volume, e *longhorn.En
 		return nil
 	}
 
-	if hardNodeAffinity != "" && getRebuildingReplicaCount(e) == 0 {
-		r, err := vc.createReplica(v, e, rs, hardNodeAffinity)
-		if err != nil {
-			return err
-		}
-		rs[r.Name] = r
+	if currentRebuilding := getRebuildingReplicaCount(e); currentRebuilding != 0 {
+		return nil
 	}
 
+	log := getLoggerForVolume(vc.logger, v)
+
 	replenishCount := vc.getReplenishReplicasCount(v, rs)
-	if len(rs) != 0 && replenishCount > 0 {
-		// For rebuild case, schedule and rebuild one replica at a time
+	// For regular rebuild case or data locality case, rebuild one replica at a time
+	if (len(rs) != 0 && replenishCount > 0) || hardNodeAffinity != "" {
 		replenishCount = 1
-		currentRebuilding := getRebuildingReplicaCount(e)
-		if currentRebuilding != 0 {
-			return nil
-		}
 	}
+
 	for i := 0; i < replenishCount; i++ {
-		r, err := vc.createReplica(v, e, rs, "")
+		hasPotentialReusableReplica, reusableFailedReplica, err := vc.scheduler.CheckAndReuseFailedReplica(rs, v, hardNodeAffinity)
```

Comment:
is `hasPotentialReusableReplica` same as `reusableFailedReplica != nil`? I don't see the difference in the flow below.

Reply:
According to Shuo
```
potential reusable replica means the failed replica may be able to reuse it later but it’s not valid now due to node/disk down issue.
```

Those are the replicas we need to wait for `ReplicaReplenishmentWaitInterval` for.

Reply:
Make sure add those to the comments.